llvm/tools/sycl-post-link/sycl-post-link.cpp

```diff
@@ -937,7 +965,83 @@ handleESIMD(module_split::ModuleDesc &&MDesc, bool &Modified,
+// For example, the intel_gpu_tgllp target does not support fp64 - therefore,
+// a module using fp64 would *not* be compatible with intel_gpu_tgllp.
+bool isTargetCompatibleWithModule(const std::optional<std::string> &Target,
+                                  module_split::ModuleDesc &IrMD) {
+  // When the user does not specify a target,
+  // (e.g. -o out.table compared to -o intel_gpu_pvc,out-pvc.table)
+  // Target will have no value and we will not want to perform any filtering, so
+  // we return true here.
+  if (!Target.has_value())
+    return true;
+
+  // TODO: If a target not found in the device config file is passed,
+  // to sycl-post-link, then we should probably throw an error. However,
+  // since not all the information for all the targets is filled out
+  // right now, we return true, having the affect that unrecognized
+  // targets have no filtering applied to them.
+  if (!is_contained(DeviceConfigFile::TargetTable, *Target))
+    return true;
+
+  const DeviceConfigFile::TargetInfo &TargetInfo =
+      DeviceConfigFile::TargetTable[*Target];
+  const SYCLDeviceRequirements &ModuleReqs =
+      IrMD.getOrComputeDeviceRequirements();
+  // The device config file data stores the target's supported
+  // aspects as a vector of the strings, so we need to translate
+  // the values to a common format.
+  const NamedMDNode *Node = IrMD.getModule().getNamedMetadata("sycl_aspects");
+  if (Node) {
+    SmallMapVector<StringRef, int, 32> AspectNameToValue;
+    for (const MDNode *N : Node->operands()) {
+      assert(N->getNumOperands() == 2 &&
+             "Each operand of sycl_aspects must be a pair.");
+
+      // The aspect's name is the first operand.
+      const auto *AspectName = cast<MDString>(N->getOperand(0));
+
+      // The aspect's integral value is the second operand.
+      const auto *AspectCAM = cast<ConstantAsMetadata>(N->getOperand(1));
+      const Constant *AspectC = AspectCAM->getValue();
+
+      AspectNameToValue[AspectName->getString()] =
+          cast<ConstantInt>(AspectC)->getSExtValue();
+    }
+
+    // Make the set of aspects values the target supports.
+    SmallSet<int64_t, 32> TargetAspectValueSet;
+    for (auto Aspect : TargetInfo.aspects) {
+      auto It = AspectNameToValue.find(Aspect);
+      assert(It != AspectNameToValue.end() && "Aspect value mapping unknown!");
+      TargetAspectValueSet.insert(It->second);
+    }
+
+    // Now check to see if all the requirements of the input module
+    // are compatbile with the target.
+    for (auto Aspect : ModuleReqs.Aspects) {
+      if (!TargetAspectValueSet.contains(Aspect))
+        return false;
+    }
+  }
+
+  // Check if module sub group size is compatible with the target.
+  if (ModuleReqs.SubGroupSize.has_value()) {
+    auto Begin = TargetInfo.subGroupSizes.begin();
+    auto End = TargetInfo.subGroupSizes.end();
+    if (std::find(Begin, End, *ModuleReqs.SubGroupSize) == End)
```

Comment:
Note: `llvm::find` should be able to just accept `TargetInfo.subGroupSizes` as-is. See LLVM `STLExtras.h` documentation

Reply:
Switched it to using `is_contained` [4d2f88d](https://github.com/intel/llvm/pull/12727/commits/4d2f88dc4fada888735b36eeae03e906587694cc)